### PR TITLE
[LOOP-349] Refactor to support service plugins

### PR DESCRIPTION
**Pull request by darinkrauss**

### CODE REVIEW

File: Loop/View Controllers/SettingsTableViewController.swift [hunk 1]
```diff
@@ -293,8 +293,8 @@ final class SettingsTableViewController: UITableViewController {
             configCell.accessoryType = .disclosureIndicator
             return configCell
         case .services:
-            if indexPath.row < servicesSorted.count {
-                let service = servicesSorted[indexPath.row]
+            if indexPath.row < activeServices.count {
+                let service = activeServicesSorted[indexPath.row]
```
Comment (nhamming) [inline Loop/View Controllers/SettingsTableViewController.swift]: Should the sorted version be used in both instances?

File: Loop/Extensions/UIAlertController.swift [hunk 2]
```diff
@@ -125,25 +125,25 @@ extension UIAlertController {
...
     ///
     /// - Parameters:
-    ///   - serviceUITypes: An array of service UI types.
-    ///   - selectionHandler: A closure to execute when a service UI is selected.
-    ///   - service: The selected service UI type.
-    internal convenience init(serviceUITypes: [ServiceUI.Type], selectionHandler: @escaping (_ serviceUIType: ServiceUI.Type) -> Void) {
+    ///   - services: An array of services.
+    ///   - selectionHandler: A closure to execute when a service is selected.
+    ///   - identifier: The identifier of the selected service.
+    internal convenience init(services: [AvailableDevice], selectionHandler: @escaping (_ identifier: String) -> Void) {
```
Comment (nhamming) [inline Loop/Extensions/UIAlertController.swift]: nit: Could just be me, but I find `services: [AvailableDevice]` confusing and I questioned if this was correct. Maybe a `Typealias` to define `AvailableService`?

File: Loop/Managers/ServicesManager.swift [hunk 3]
```diff
@@ -7,53 +7,114 @@
...
+    private let servicesLock = UnfairLock()
 
     private var observers = WeakSet<ServicesManagerObserver>()
 
-    var services: [Service] {
-        didSet {
-            dispatchPrecondition(condition: .onQueue(.main))
-            UserDefaults.appGroup?.services = services
-            notifyObservers()
+    private let observersLock = UnfairLock()
```
Comment (nhamming) [inline Loop/Managers/ServicesManager.swift]: Interesting that the observers are also locked. Should this be common practice? I have noticed other places (e.g., DexcomCGMKit) where this isn't the case.

File: Loop/View Controllers/SettingsTableViewController.swift [hunk 4]
```diff
@@ -721,35 +718,47 @@ extension SettingsTableViewController: CGMManagerSetupViewControllerDelegate {
 }
 
 extension SettingsTableViewController: ServiceSetupDelegate {
-    fileprivate var serviceTypesAvailable: [Service.Type] {
-        return serviceTypes.filter { serviceType in !dataManager.servicesManager.services.contains { type(of: $0) == serviceType } }
+    fileprivate var availableServices: [AvailableDevice] {
+        return dataManager.servicesManager.availableServices
+    }
```
Comment (nhamming) [inline Loop/View Controllers/SettingsTableViewController.swift]: Similar nit here about `availableServices` defined as `availableDevice`

**Review by nhamming — APPROVED** [review]
Just general comments/questions.

File: Loop/View Controllers/SettingsTableViewController.swift (hunk 1 above)
Comment (darinkrauss) [reply Loop/View Controllers/SettingsTableViewController.swift]: In the two cases where `activeServices` is used, it actually doesn't matter, but for clarity, yes, I will update.

File: Loop/Extensions/UIAlertController.swift (hunk 2 above)
Comment (darinkrauss) [reply Loop/Extensions/UIAlertController.swift]: 👍

File: Loop/Managers/ServicesManager.swift (hunk 3 above)
Comment (darinkrauss) [reply Loop/Managers/ServicesManager.swift]: I'd think so.

**Review by nhamming — APPROVED** [review]
LGTM

File: Loop.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -71,7 +71,6 @@
 		4345E40421F68AD9009E00E5 /* TextRowController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4345E40321F68AD9009E00E5 /* TextRowController.swift */; };
 		4345E40621F68E18009E00E5 /* CarbEntryListController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4345E40521F68E18009E00E5 /* CarbEntryListController.swift */; };
 		4346D1E71C77F5FE00ABAFE3 /* ChartTableViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4346D1E61C77F5FE00ABAFE3 /* ChartTableViewCell.swift */; };
-		434FB6461D68F1CD007B9C70 /* Amplitude.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 434FB6451D68F1CD007B9C70 /* Amplitude.framework */; };
```
Comment (ps2) [inline Loop.xcodeproj/project.pbxproj:74]: Love to see all these files drop out of Loop!

File: Loop/AppDelegate.swift [hunk 6]
```diff
@@ -17,13 +17,15 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
 
     private lazy var log = DiagnosticLog(category: "AppDelegate")
 
-    private lazy var servicesManager = ServicesManager()
+    private lazy var pluginManager = PluginManager()
+
+    private lazy var servicesManager = ServicesManager(pluginManager: pluginManager)
 
     private lazy var analyticsServicesManager = AnalyticsServicesManager(servicesManager: servicesManager)
```
Comment (ps2) [inline Loop/AppDelegate.swift]: This seems inverted. If the ServicesManager owns the other managers, then you can use an imperative style of programming that should save on locks, and replace some of the observe/notify style interactions that are on the ServicesManager with direct calls. I think at this layer of Loop, direct control is probably more appropriate than loosely coupling things.

File: Loop/Plugins/LoopPlugins.swift [hunk 7]
```diff
@@ -102,13 +104,51 @@ class PluginManager {
...
+
 }
 
 
 extension Bundle {
     var isLoopPlugin: Bool {
         return
             object(forInfoDictionaryKey: LoopPluginBundleKey.pumpManagerIdentifier.rawValue) as? String != nil ||
-            object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerIdentifier.rawValue) as? String != nil
+            object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerIdentifier.rawValue) as? String != nil ||
```
Comment (ps2) [inline Loop/Plugins/LoopPlugins.swift:151]: I think having to open up *all* the framework's plist files is going to be felt as slow by users. My original idea was to name the plugins with a .loopplugin extension, but that doesn't get past app store rules.  I'm now thinking plugin bundles should either contain a special file that denotes them as plugins, or the plugin installer script generates a manifest.  Leaning toward the second idea. No changes needed here; just thinking out loud. :)

**Review by ps2 — COMMENTED** [review]
I would consider restructuring the relationship between `ServicesManager` and the subtype manager instances, to simplify the interactions.  I'm not sure the observer pattern (and its associated locking) is needed. If ServicesManager owns and is responsible for the subtypes, they probably don't need to know how they're being managed, and can be controlled directly.

**Review by nhamming — APPROVED** [review]
+1

File: Loop/Plugins/LoopPlugins.swift (hunk 7 above)
Comment (darinkrauss) [reply Loop/Plugins/LoopPlugins.swift:151]: :+1: for the second idea.

File: Loop/AppDelegate.swift (hunk 6 above)
Comment (darinkrauss) [reply Loop/AppDelegate.swift]: I believe I have address your concerns here in the latest go around. Please review and let me know what you think.

**Review by ps2 — APPROVED** [review]
LGTM

**Comment by darinkrauss** [comment]
@ps2 Last commit breaks apart the service manager again per your PR review thoughts.

**Review by ps2 — APPROVED** [review]
LGTM!

**Pull request merged**